src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,323 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import tactic.basic
+import field_theory.finite.galois_field
+import number_theory.cyclotomic.primitive_roots
+import field_theory.finite.trace
+
+/-!
+# Additive characters of finite rings and fields
+
+Let `R` be a finite commutative ring. An *additive character* of `R` with values
+in another commutative ring `R'` is simply a morphism from the additive group
+of `R` into the multiplicative monoid of `R'`.
+
+We use the namespace `add_char`.
+
+## Main definitions and results
+
+We define `mul_shift ψ a`, where `ψ : add_char R R'` and `a : R`, to be the
+character defined by `x ↦ ψ (a * x)`. An additive character `ψ` is *primitive*
+if `mul_shift ψ a` is trivial only when `a = 0`.
+
+We show that when `ψ` is primitive, then the map `a ↦ mul_shift ψ a` is injective
+(`add_char.to_mul_shift_inj_of_is_primitive`) and that `ψ` is primitive when `R` is a field
+and `ψ` is nontrivial (`add_char.is_primitive_of_is_nontrivial`).
+
+We also show that there are primitive additive characters on `R` (with suitable
+target `R'`) when `R` is a field or `R = zmod n` (`add_char.primitive_char_finite_field`
+and `add_char.primitive_zmod_char`).
+
+Finally, we show that the sum of all character values is zero when the character
+is nontrivial (and the target is a domain); see `add_char.sum_eq_zero_of_is_nontrivial`.
+
+## Tags
+
+additive character
+-/
+
+/-!
+### Definitions related to and results on additive characters
+-/
+
+section add_char_def
+
+universes u v
+
+-- The domain of our additive characters
+variables (R : Type u) [comm_ring R]
+-- The target
+variables (R' : Type v) [comm_ring R']
+
+/-- Define `add_char R R'` as `(multiplicative R) →* R'`. -/
+abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+
+end add_char_def
+
+
+section additive
+
+namespace add_char
+
+open multiplicative
+
+universes u v
+
+-- The domain of our additive characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial additive character; it sends everything to `1`. -/
+@[simps] protected
+def trivial : add_char R R' :=
+{ to_fun := λ x, 1,
+  map_one' := rfl,
+  map_mul' := λ x y, (mul_one 1).symm }
+
+/-- An additive character is *nontrivial* if it takes a value `≠ 1`. -/
+def is_nontrivial (ψ : add_char R R') : Prop := ∃ (a : R), ψ a ≠ 1
```

Comment:
```suggestion
def is_nontrivial (ψ : add_char R R') : Prop := ∃ (a : R), ψ (of_add a) ≠ 1
```

Reply:
Is `∃ (a : R), ψ (of_add a) ≠ 1` preferable over `∃ (a : multiplicative R), ψ a ≠ 1`?

Reply:
It looks like your version leads to fewer changes down the line, so I'll use yours.